services/fission-user-workers/service/ctrl_docker.go

```diff
@@ -0,0 +1,252 @@
+			log.Printf("Error removing container: %s: %v\n", c.ID, err)
+			continue
+		}
+	}
+
+	// Iterate over workers in Database
+	pages, err := ctrl.store.ListUserWorkers(ListWorkerFilters{State: StateEnabled}, pagination.Request{Limit: 10})
+	if err != nil {
+		return fmt.Errorf("error listing user workers from database: %w", err)
+	}
+	for {
+		toCreate := []UserWorker{}
+		toUpdate := []UserWorker{}
+		for _, worker := range pages.Data {
+			if worker.Language != LanguagePython {
+				continue
+			}
+			filter := filters.NewArgs(
+				filters.Arg("label", "controlled-by=sensorbucket"),
+				filters.Arg("label", "worker-id="+worker.ID.String()),
+			)
+			candidates, err := ctrl.docker.ContainerList(ctx, types.ContainerListOptions{
+				All:     true,
+				Filters: filter,
+			})
+			if err != nil {
+				log.Printf("Error getting containers for worker-id: %s: %v\n", worker.ID.String(), err)
+				continue
+			}
+			// Must be created
+			if len(candidates) == 0 {
+				log.Printf("WHY: creating %s because it has no instances\n", worker.ID.String())
+				toCreate = append(toCreate, worker)
+				continue
+			} else if len(candidates) > 1 {
+				// Updating is deleting then creating, so this will remove all duplicates and then create a single
+				log.Printf("WHY: updating %s because it has too many instances\n", worker.ID.String())
+				toUpdate = append(toUpdate, worker)
+				continue
+			}
+			container := candidates[0]
+			revision, err := strconv.ParseInt(container.Labels["worker-revision"], 10, 64)
+			if err != nil {
+				log.Printf("Error converting worker revision to int for container: %s\n", container.ID)
+				continue
+			}
+			if revision < int64(worker.Revision) {
+				log.Printf("WHY: updating %s because it's revision is lower (%d < %d)\n", worker.ID.String(), revision, worker.Revision)
+				toUpdate = append(toUpdate, worker)
+				continue
+			}
+			if !lo.Contains([]string{"running", "starting"}, container.State) {
+				log.Printf("WHY: updating %s because it's state dictates so: %s\n", worker.ID.String(), container.State)
+				toUpdate = append(toUpdate, worker)
+			}
+		}
+
+		// Commit changes
+		log.Printf("Creating %d workers...\n", len(toCreate))
+		for _, worker := range toCreate {
+			if err := ctrl.createContainerForWorker(ctx, worker); err != nil {
+				log.Printf("Error creating container for worker: %v\n", err)
+			}
+		}
+		log.Printf("Updating %d workers...\n", len(toUpdate))
+		for _, worker := range toUpdate {
+			if err := ctrl.removeContainer(ctx, worker); err != nil {
+				log.Printf("Error removing container for worker: %v\n", err)
+				continue
+			}
+			if err := ctrl.createContainerForWorker(ctx, worker); err != nil {
+				log.Printf("Error creating container for worker: %v\n", err)
+			}
+		}
+
+		// Continue to next page if there is one
+		if pages.Cursor == "" {
+			break
+		}
+		pages, err = ctrl.store.ListUserWorkers(ListWorkerFilters{State: StateEnabled}, pagination.Request{Cursor: pages.Cursor})
```

Comment:
maybe just move this call to the start of the for loop and declare the cursor outside of the loop. then you can remove the List workers call from before the start of the loop